src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs

```diff
@@ -386,6 +386,252 @@ IValue CopyAndUpdateUdt(TupleValue originalValue)
+                        var isLast = idx == tupleElements.Length - 1;
+                        var next = isLast
+                            ? CreateConstantString(")")
+                            : comma ?? CreateConstantString(",");
+                        str = DoAppend(str, next, unreferenceNext: isLast);
+                    }
+
+                    if (comma != null)
+                    {
+                        UpdateStringRefCount(comma, -1);
+                    }
+                    return str;
+                }
+
+                var ty = evaluated.QSharpType.Resolution;
+                if (ty.IsString)
+                {
+                    UpdateStringRefCount(evaluated.Value, 1);
+                    return evaluated.Value;
+                }
+                else if (ty.IsBigInt)
+                {
+                    return SimpleToString(RuntimeLibrary.BigIntToString);
+                }
+                else if (ty.IsBool)
+                {
+                    return SimpleToString(RuntimeLibrary.BoolToString);
+                }
+                else if (ty.IsInt)
+                {
+                    return SimpleToString(RuntimeLibrary.IntToString);
+                }
+                else if (ty.IsResult)
+                {
+                    return SimpleToString(RuntimeLibrary.ResultToString);
+                }
+                else if (ty.IsPauli)
+                {
+                    return SimpleToString(RuntimeLibrary.PauliToString);
+                }
+                else if (ty.IsQubit)
+                {
+                    return SimpleToString(RuntimeLibrary.QubitToString);
+                }
+                else if (ty.IsRange)
+                {
+                    return SimpleToString(RuntimeLibrary.RangeToString);
+                }
+                else if (ty.IsDouble)
+                {
+                    return SimpleToString(RuntimeLibrary.DoubleToString);
+                }
+                else if (ty.IsFunction)
+                {
+                    return CreateConstantString("<function>");
+                }
+                else if (ty.IsOperation)
+                {
+                    return CreateConstantString("<operation>");
+                }
+                else if (ty.IsUnitType)
+                {
+                    return CreateConstantString("()");
+                }
+                else if (ty.IsArrayType)
+                {
+                    var arr = (ArrayValue)evaluated;
+                    var str = CreateConstantString("[");
+                    Value comma = CreateConstantString(",");
+                    Value closeParens = CreateConstantString("]");
+
+                    sharedState.IterateThroughArray(arr, item =>
+                    {
+                        str = DoAppend(str, ExpressionToString(item));
+                        str = DoAppend(str, comma, unreferenceNext: false);
+                    });
```

Comment:
I have an impression that this fragment inserts comma `,` after the last element and before `]`.
If that happens, then just in case I wanted to make sure that such an insertion is OK.

Reply:
It does. I was on the fence about this. I could generate a conditional branch to not do that, but thinking of automatic parsing tools and that it is not uncommon for larger sequences to list them across multiple lines with each item being followed by a comma including the last one (making it easier to append), I opted for going with the more readable version of just leaving it there.

Reply:
I typically prefer to handle the first element separately.
And then, if there are more elements, add comma and one more element, comma and one more element, .. on each new iteration.
(if there is a chance that comma after the last element can cause a problem)